### PR TITLE
Journalist can create and modify article

### PR DESCRIPTION
**Pull request by piavW**
[PT-link](https://www.pivotaltracker.com/story/show/169282741)

## PR Content
Creates functionality for an editor to give a journalist access to create an article (makes it possible to create an article). 

## This PR Contains
- Unit tests of user model
- Pundit gem
- Create action for article model 
- Possibility to attach image to article
- Roles attributes to user model
- Create & edit action for user model
- Request tests for article model create, update, index actions
- Request test for authorization of create and update actions

## What we've learned
- How to use pundit and pundit-matchers
- How to set a policy and test that policy
- How to create associations with a role of a user model

### CODE REVIEW

File: app/controllers/api/v1/articles_controller.rb [hunk 1]
```diff
@@ -1,4 +1,5 @@
 class Api::V1::ArticlesController < ApplicationController
+  before_action :authenticate_user!
```
Comment (DIzosimov) [inline app/controllers/api/v1/articles_controller.rb]: space between before action and index

**Review by miyeaier — APPROVED** [review]

File: spec/models/article_spec.rb [hunk 2]
```diff
@@ -16,7 +16,11 @@
 
   describe 'Factory' do
     it 'should have valid Factory' do
-      expect(create(:article)).to be_valid 
+      expect( create(:article)).to be_valid
```
Comment (DIzosimov) [inline spec/models/article_spec.rb]: random space before create

**Review by DIzosimov — CHANGES_REQUESTED** [review]
check comment (random space thats need to be removed)

**Review by yasmineezequiel — APPROVED** [review]

**Review by beccaburns — APPROVED** [review]

File: config/routes.rb [hunk 3]
```diff
@@ -1,12 +1,14 @@
...
-  get 'articles/index'
   namespace :api do
     namespace :v1 do
-      resources :articles, only: [:index], constraints: { format: 'json' }
+      resources :articles, only: [:index, :create, :update], constraints: { format: 'json' }
     end
 
     namespace :v1 do
-      mount_devise_token_auth_for 'User', at: 'auth', skip: [:omniauth_callbacks]
+      # mount_devise_token_auth_for 'User', at: 'auth', skip: [:omniauth_callbacks]
```
Comment (faraznaeem) [inline config/routes.rb]: Please clean out the comments

File: spec/models/article_spec.rb [hunk 4]
```diff
@@ -19,4 +19,8 @@
       expect(create(:article)).to be_valid 
     end
   end
+
```
Comment (faraznaeem) [inline spec/models/article_spec.rb]: Add a test for the attached image

**Review by faraznaeem — CHANGES_REQUESTED** [review]

File: app/controllers/api/v1/articles_controller.rb [hunk 5]
```diff
@@ -9,9 +11,48 @@ def index
...
+  def show
+    @article = Article.find(params[:id])
+  end
+
+  def edit
+    @article = Article.find(params[:id])
+  end
+
+  def update
+    @article = Article.find(params[:id])
```
Comment (tochman) [inline app/controllers/api/v1/articles_controller.rb:31]: You need to authorize the user for `update` action

File: app/policies/article_policy.rb [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  def index?
+    true
+  end
+
+  def show?
+    true
+  end
+
+  def new?
+    @user.journalist?
```
Comment (tochman) [inline app/policies/article_policy.rb:17]: There is no `new` actions in API's

File: spec/models/article_spec.rb [hunk 7]
```diff
@@ -19,4 +19,19 @@
       expect(create(:article)).to be_valid 
     end
   end
+  
+  it 'Should have a journalist' do
+    should belong_to(:journalist)
```
Comment (tochman) [inline spec/models/article_spec.rb]: ?? 😳

File: spec/policies/article_policy_spec.rb [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  context 'user of role subscriber cannot edit or update an article' do
+    let(:user) { User.create }
+
+    it { is_expected.to forbid_edit_and_update_actions }
+  end
+end
+
+describe ArticlePolicy do
+  subject { described_class.new(user, article) }
+  let(:user) { User.create(role: 'journalist') }
```
Comment (tochman) [inline spec/policies/article_policy_spec.rb]: Please use factories when creating objects in tests

**Review by tochman — CHANGES_REQUESTED** [review]
There are issues on this one

File: spec/models/article_spec.rb (hunk 7 above)
Comment (piavW) [reply spec/models/article_spec.rb]: Yea that test isn't doing anything really is it...

**Review by DIzosimov — APPROVED** [review]

**Review by miyeaier — APPROVED** [review]

File: spec/policies/article_policy_spec.rb (hunk 8 above)
Comment (faraznaeem) [reply spec/policies/article_policy_spec.rb]: Please address this

Comment (piavW) [reply spec/policies/article_policy_spec.rb]: my bad, missed that one when I changed the other ones.

**Review by faraznaeem — APPROVED** [review]

**Pull request merged**